Basic Tidal Mini Notation Parser

unlessgames opened the pull request:
Here is the first draft. Please excuse my rust, I'm sure I am not using the proper rust idioms, features and what-not, any hints would be welcome.

I've implemented the basics of the mini notation (excluding any operators for now) by building an intermediate recursive structure `Step`. The pest AST is converted into this form then `output_events` can generate a flat list of events with start and end times. 

```bash
cargo test --lib -- --exact rhythm::tidal::test::test_tidal --show-output
```
This will run the test and print the AST, the parsed structure and output events for a bunch of test strings. 

Step literals are parsed as they should be
* `Integer`s as expected `3`, `5`, `8`, `-13`
* `Float`s can be typed as `1.0`, `.5` or `1.`
* `Pitch` is parsed from case-insensitive notes from `abcdefg` with an optional sharp/flat mark and octave like `a#3` or `Ab4`
* arbitrary string identifiers like `kick`, `lick` or whatever, turn into `Name`
*  `~` as `Rest`
* `_` as `Hold`

Pattern groupings are mostly working, including nested structures
* `[ ]` as `Subdivision` (which is what the root step of any cycle is as well)
* `< >` as `Alternating`, these keep their state as `current` index
* `{ }%N` as `Polymeter` with inner state `offset`. The variant in Tidal without `%N` parameter is not supported, I think that is kind of confusing anyway since you have to remember which side wraps the other and it won't do anything without an inner `,`
* `|` to choose randomly from a set works can only be used inside subdivisions (just like in Tidal)
* `,` to stack patterns somewhat works

What isn't implemented:

The shorthand for subdivisions like `a b . c d` -> `[a b] [c d]`, would be nice to have but not really a priority imo

Operator expressions
Probably doable if I ever get around implementing them
* `*` to repeat a pattern
* `?` to mute steps randomly
* `:` to select instrument
* `(s,p,r)` for Euclidean sequences, it's basically just a type of subdivision

The ones that are a bit harder to do because of how I implemented everything to be a step, these would need to add steps instead of adding a Subdivision or changing existing events, it's simpler if the right hand-side is constant but I think it should be any other pattern like in tidal. I'll have to think more about how to cover these nicely..
* `/` to slow down a pattern, this is somewhat unpredictable to use in Tidal
* `@` to elongate a pattern aka `a@3 b`  -> `a _ _ b`
* `!` to replicate a pattern aka `a!3 b` -> `a a a b`

What definitely needs work in the current feature set
* `_` and `~` should be incorporated/filtered out from the outputs, currently they just show up as Hold/Rest events
* error handling and reporting should be cleaned up and made safer, I suppose no `unreachable!` should remain even if the parser grammar seemingly makes something impossible.
* stacks currently just push stuff into the event list, resulting in an unsorted list, it would be probably better to output to a separate vec with each stack and zip them together at the end.
* there should be a converter to eventually get the right type of events instead of the current internal CycleEvent

unlessgames commented:
I reworked the code to support different lengths for each step, to prepare for the eventual `!`, `@` operators. Unfortunately I'm not sure how to make it support `/`. As it behaves in a way that would need us to allow arbitrary time interval querying at every level, for example in Tidal writing `a/2 b` would result in `a b` in the first cycle and `~ b` in the second, as if the `a` pattern was running at half speed across cycles. It's a bit counter-intuitive to use in my experience so it might not be too bad to omit it from here (`<>` is often a better way to do such things), the lengthening operators however would be good to have and they should be possible now.

I also changed how stacks are generated with the Events enum (renames might be good here), they output to separate channels now so writing the final conversion to afseq events should be nicer.

Fixed a few things with the grammar, I think there are some issues with trailing characters that we will have to test more.

Applied some of the suggestions you said, I still have to fix the random generator and make the Parser more self contained.

When it comes to testing, we are going to need a bit more complexity with asserts to properly check the output of random generations.

unlessgames commented:
Alternatively we could implement `/` as `a/3 b` -> `a ~ ~ b` which is somewhat of a missing shorthand from Tidal itself.

Speaking of alternatives, I think it would be nice to add an operator for random choice from a `[ ]` subdivision. For example `[a b c]%` -> `[a | b | c]` (or even just `(a b c)` although that clashes a bit with euclidean notation), because having to write `|` between each choice is not the most ergonomic for playing around.

emuell commented:
> for example in Tidal writing a/2 b would result in a b in the first cycle and ~ b in the second, as if the a pattern was running at half speed across cycles. It's a bit counter-intuitive to use in my experience so it might not be too bad to omit it from here (<> is often a better way to do such things), the lengthening operators however would be good to have and they should be possible now.

Personally, I can't get my head around it either, so skipping it seems like a good idea for now. Especially in the first iteration,
and if the parser spits out a pretty clear error (not supported, blah).

 >  I still have to fix the random generator and make the Parser more self contained.
 
 As mentioned before, I'm happy to help out with the plain rust stuff, if that helps you to concentrate on the other things.
 
 > Alternatively we could implement / as a/3 b -> a ~ ~ b which is somewhat of a missing shorthand from Tidal itself.
 
But then we end up writing our own dialect of notation? I would avoid that because it's confusing for people coming from tidal. And even for new people, it's a lot easier to point them to some tutorials for tidal than to reinvent the wheel.  So I think it's better to skip things than to reinterpret them.

unlessgames commented:
You are right about the "dialect". I guess adding new stuff that means nothing or invalid in Tidal could be fine, for example I currently treat empty `[]` groups as an empty step `~` where in Tidal that simply won't parse.

---

Now I've implemented some expression operators, although they only work with single values on the right hand side. Properly supporting everything on the right is less trivial. For example in Tidal `[a b]*[2 3]` returns `[a b][b a b]` as if both `2` and `3` generated a full cycle (`[a b a b]` and `[a b a b a b]` respectively) and the pattern switched between them half-way through.

So only single values on the right for now and with that these are working currently
* `*` repeats the step multiple times inside the same interval
* `:` sets the `target` field on the outgoing events (this can be either `None`, and integer `Index` (like an instrument number for example) or a string `Name`, this latter might not be needed (and it doesn't exist in Tidal) but for remapping purposes it might be useful to have here)
* `?` uses the random generator to switch values to an empty step on occasion
* `!` duplicates the step n times taking up more relative space inside the pattern
* `(p,s,r)` repeats the left hand side according to the Bjorklund pattern with pulses, steps and an optional rotation value

What remains
* `/` as discussed above
* `@` which is giving me a bit of trouble still
* `..` range operator, kinda forgot about that before..

Overall they seem to work from simple tests but a bit more AB testing is required as I have doubts about whether or not they behave the same way as in Tidal for all sorts of nested cases (the `!` operator in particular).

unlessgames commented:
Ok, I think I'll stop here as far as features go and leave the remaining operators and supporting nested stuff on the right hand side for the future, so we can focus on integration and making the existing code better.

Got rid of `unreachable!()` to return `Err(String)` everywhere with some info attached. 

Moved the parsing, the random generator and output into a Cycle struct, now generating output from a string looks like this

```rust
//                           seed
match Cycle::from("a b c d", None) {
  Ok(mut cycle) => {
    let mut channels: Vec<Vec<SingleEvent>> = Cycle::generate(&mut cycle);
  Err(err) => () // the parsing failed
}
```

The `channels` each contain a monophonic list of events with `StepValue`s. Handling of rests, holds and conversion to shared types is still needed to be done. I guess we have to consider that

* Channels here can have arbitrary event boundaries, on one channel an event might start at 0.0 and end at 0.5 while on the other channel an event might also start at 0.0 but end at 0.25. This complicates batching into a single NoteEvents because of the different note off times needed. As far as I understand, afseq currently has no notion of note off events, right?
* Channels can contain different type of events on a single channel or across multiple ones for example `[a b, 0.5 0.8]` will lead to one channel having notes and the other floats or `a b 0.5 0.8` will lead to a mixed stream of events.

unlessgames commented:
I started writing tests (after adding your patch) but it seems floating point errors will make this very annoying using f64, once we try to cover spans with recurring decimals and such.

I added the [fraction crate](https://docs.rs/fraction/latest/fraction/) to the project as a test, it works great and it was very simple to switch all of the f64 to using fractions, writing tests can also be done much nicer now. 

As it adds a new dependency your approval is definitely needed here before going further. Of course we would still convert to f64 or whatever is needed in the end but since everything is essentially fractions here it seems more than fitting to use this on the "Tidal side" I think.

emuell commented:
Looks good to me. I would not have used the rather cryptic Shortcut F for it, but just Fraction, else probably would have done exactly the same. 

Yet another good thing in Rust, that such dependencies are easily reachable and that most rust modules are really well documented and maintained as well.

unlessgames commented:
Converted the debug prints to tests and fixed two bugs in the generation code in the meantime. The code now uses "Fraction" instead of "F" (except in tests).

Custom tests for random choices and operators are still missing.

unlessgames commented:
Added some helpers to `Events` that handle merging the output. `Events::merge` removes `Hold`s while extending the previous event (if any) then it removes any consecutive `Rest`s. Any `Rest` that remains is essentially an OFF event for the preceding event. This should make conversion easier.

unlessgames commented:
With that, I think it can be merged now if it's alright.

emuell commented:
Looks and works great. If you're happy with it, it's done. 

The only final suggestions I'd have are to fix the [clippy warnings](https://doc.rust-lang.org/stable/clippy/usage.html) and remove dead commented out code: the project is configured to use clippy. Clippy can be annoying as hell, but in general helps a lot to find suspicious code and streamline it using common Rust practices. 

If you're using vscode, you can enable it [like this](https://users.rust-lang.org/t/how-to-use-clippy-in-vs-code-with-rust-analyzer/41881).

Attached is a patch that fixes the warnings, but please feel free to tweak the changes to your liking. 
[Clippy.patch](https://github.com/emuell/afseq/files/15286102/Clippy.patch)

--- 

I will then start next week with the integration by:

- either creating a new CycleRhythm impl which takes a Cycle as argument in Rust `cycle_time(1234).trigger("a b d")` or extending the existing "Generic" rhythms to optionally use a cycle as **emitter** only and use the existing beat and second times `beat_time.with_pattern([1]).trigger_cycle("a b d")`. I do prefer the latter, because it means less duplicated code and more options and better integration into what's already there, but need to see how this goes while doing it. We have a few more options in Lua to streamline the integration.
- then adding Lua bindings for it (just the basic stuff)
- then allowing to map `Target` to note values (or chords or whatever else)

For the Rhythm integration I likely need to change parts of the interfaces, make some stuff pub(crate) others private. Maybe also will use `Note` directly in the Cycle, but a conversion from Pitch to Note could work well too.

I'll open pull requests for all this, so you can jump in and tweak stuff too.

---

Do you have a list of open/planed things/TODOs? Then let's add a new issue for that or move them somewhere else?

unlessgames commented:
Neat, clippy has some nice suggestions! I've fixed the warnings and added some `#[allow(clippy..]` for the places where I want to keep the match statements for possible future extensions.

I removed the commented out code from the tests.

Saved the input string to the cycle and added a function that can be used to check if a Cycle is stateful, if not, every iteration will produce the same output.

I can't say I completely understand the implications of the options for integration you've listed (definitely still need to play around with the rest of the codebase a bit more), but the second one does sound better.

> Maybe also will use `Note` directly in the Cycle, but a conversion from Pitch to Note could work well too.

I didn't change to the built-in `Note` because I liked it better to have the whole tidal part be more self-contained and because the Note is a bit different in that it includes the OFF and EMPTY values which are distinct types here. Since a conversion is easy to write I think it would be better to have that as an explicit point of separation instead.

> Do you have a list of open/planed things/TODOs?

I left some `TODO` comments in the code, but it mostly boils down to

* allow any group on the right side of Expressions, inside Bjorklund and as count for Polymeters, like `kick*<1 2 3>` or `[a b]*[2 3]`. This requires a fair bit of work.
* the remaining operators `@` and `/`, I have a feeling that supporting the above will make these easier as well
* the range expressions like `1..4` -> `1 2 3 4`
* the shorthand for `Subdivision` like `a b . c d` -> `[a b] [c d]`
* some extra non-existent random operators, if allowed
    * a short-hand for Choice (for example `[1 2 3]#` -> `[1 | 2 | 3]`)
    * something that will shuffle the insides of any group on each iteration (`[1 2 3]$` -> `[2 1 3]; [3 2 1]; [3 1 2]...`)

emuell commented:
Alright. Let's consider this done and move on then.
